### PR TITLE
fix(request-id): file descriptor overflow when use request-id plugin nanoid algorithm

### PR DESCRIPTION
**Pull request by aikin-vip**
### Description

Fixes  (#8931)

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: t/plugin/request-id3.t [hunk 1]
```diff
@@ -0,0 +1,186 @@
+#
```
Comment (zll600) [inline t/plugin/request-id3.t:1]: what about `requiest-id-nanoid.t`?

Comment (aikin-vip) [reply t/plugin/request-id3.t:1]: sure

File: docs/en/latest/plugins/request-id.md [hunk 2]
```diff
@@ -47,6 +47,8 @@ The Plugin will not add a unique ID if the request already has a header with the
 | algorithm           | string  | False    | "uuid"         | ["uuid", "nanoid", "range_id"] | Algorithm to use for generating the unique request ID.                 |
 | range_id.char_set      | string | False | "abcdefghijklmnopqrstuvwxyzABCDEFGHIGKLMNOPQRSTUVWXYZ0123456789| The minimum string length is 6 | Character set for range_id |
 | range_id.length    | integer | False | 16             | Minimum 6 | Id length for range_id algorithm |
+| nanoid.char_set      | string | False | "abcdefghijklmnopqrstuvwxyzABCDEFGHIGKLMNOPQRSTUVWXYZ0123456789_-| The minimum string length is 6 | Character set for nanoid |
+| nanoid.length    | integer | False | 21             | Minimum 6 | Id length for range_id algorithm |
```
Comment (liweitianux) [inline docs/en/latest/plugins/request-id.md:51]: Typo. The description should be `Id length for nanoid algorithm`.

Comment (aikin-vip) [reply docs/en/latest/plugins/request-id.md]: You are right, done

File: apisix/plugins/request-id.lua [hunk 3]
```diff
@@ -51,6 +51,22 @@ local schema = {
                 }
             },
             default = {}
+        },
+        nanoid = {
```
Comment (shreemaan-abhishek) [inline apisix/plugins/request-id.lua:55]: why is this block needed when just replacing the dependency would fix the issue?

File: apisix/plugins/request-id.lua [hunk 4]
```diff
@@ -79,8 +95,9 @@ local function get_request_id(conf)
     if conf.algorithm == "uuid" then
         return uuid()
     end
+
     if conf.algorithm == "nanoid" then
-        return nanoid.safe_simple()
+        return nanoid.generate(conf.nanoid.length,conf.nanoid.char_set)
```
Comment (shreemaan-abhishek) [inline apisix/plugins/request-id.lua:100]: space needed after comma

File: t/plugin/request-id.t [hunk 5]
```diff
@@ -437,71 +437,3 @@ GET /opentracing
...
-            local t = require("lib.test_admin").test
-            local http = require "resty.http"
-            local v = {}
-            local ids = {}
-            local code, body = t('/apisix/admin/routes/1',
-                 ngx.HTTP_PUT,
-                 [[{
-                        "plugins": {
-                            "request-id": {
-                                "algorithm": "nanoid"
```
Comment (shreemaan-abhishek) [inline t/plugin/request-id.t:456]: your PR suggests that usage method/configuration for nanoid has changed. Which means existing users will face incompatibility issues when upgrading to the newer version. We should avoid such a situation.

File: apisix/plugins/request-id.lua (hunk 3 above)
Comment (aikin-vip) [reply apisix/plugins/request-id.lua:55]: The new dependency provides more controllable parameters and more customized functions, so parameter options are added，just like range_id

File: apisix/plugins/request-id.lua (hunk 4 above)
Comment (aikin-vip) [reply apisix/plugins/request-id.lua:100]: > space needed after comma

OK

File: t/plugin/request-id.t (hunk 5 above)
Comment (aikin-vip) [reply t/plugin/request-id.t:456]: I have noticed the question you raised before submitting. I have already set the initial default value in the parameters, the default parameter value will be the same as the default value generated in previous versions without modifying the parameter value. Therefore, I think the upgrade should be smooth. Is there anything missing from me?

File: t/plugin/request-id-nanoid.t [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        }
+    }
+--- response_body
+passed
+
+
+
+=== TEST 3: hit
+--- request
+GET /opentracing
```
Comment (shreemaan-abhishek) [inline t/plugin/request-id-nanoid.t:108]: add error log and status code checks here.

**Comment by Yokutto** [comment]
Uh... I think we should re-open this

**Pull request closed without merging**